okio-zipfilesystem/src/jvmMain/kotlin/okio/zipfilesystem/zip.kt

```diff
@@ -175,22 +175,60 @@ fun open(zipPath: Path, fileSystem: FileSystem): ZipFileSystem {
   // public API doesn't allow us to throw IOException except from the constructor
   // or from getInputStream.
   cursor.seek(record.centralDirOffset)
-  val entries = mutableMapOf<Path, ZipEntry>()
+  val entries = mutableListOf<ZipEntry>()
   for (i in 0 until record.numEntries) {
     val newEntry = source.readEntry(
       isZip64 = zip64EocdRecordOffset != -1L
     )
     if (newEntry.localHeaderRelOffset >= record.centralDirOffset) {
       throw IOException("Local file header offset is after central directory")
     }
-    val entryName = newEntry.name
-    val path = "/".toPath().div(entryName)
-    if (entries.put(path, newEntry) != null) {
-      throw IOException("Duplicate path: $entryName")
+    entries += newEntry
+  }
+
+  val index = buildIndex(entries)
+
+  return ZipFileSystem(zipPath, fileSystem, index, comment)
+}
+
+/**
+ * Returns a map containing all of [entries], plus parent entries required so that all entries
+ * (other than the file system root `/`) have a parent.
+ */
+@ExperimentalFileSystem
+private fun buildIndex(entries: List<ZipEntry>): Map<Path, ZipEntry> {
+  val result = mutableMapOf<Path, ZipEntry>()
+
+  // Iterate in sorted order so each path is preceded by its parent.
+  for (entry in entries.sortedBy { it.canonicalPath }) {
```

Comment:
Sort based on the number of slashes instead? depth will guarantee the order you went and this likely reduces the amount to sort?

Reply:
I don’t think the additional performance is worth the additional code.